website/docs/r/cloud_provider_access.markdown

```diff
@@ -0,0 +1,68 @@
+---
+layout: "mongodbatlas"
+page_title: "MongoDB Atlas: mongodbatlas_cloud_provider_access"
+sidebar_current: "docs-mongodbatlas-datasource-cloud-provider-access"
+description: |-
+    Provides a Cloud Provider Access settings resource register, authorization, and deauthorization
+---
+
+# mongodbatlas_cloud_provider_access
+
+`mongodbatlas_cloud_provider_access` Allows you to register and authorize AWS IAM roles in Atlas.
+
+-> **NOTE:** Groups and projects are synonymous terms. You may find `groupId` in the official documentation.
+
+-> **NOTE:** The update of the argument iam_assumed_role_arn is one step in a procedure to create unified AWS access for Atlas services. For the complete procedure, see [Set Up Unified AWS Access](https://docs.atlas.mongodb.com/security/set-up-unified-aws-access/#set-up-unified-aws-access).
+
+## Example Usage
+
+```hcl
+
+resource "mongodbatlas_cloud_provider_access" "test_role" {
+   project_id = "<PROJECT-ID>"
+   provider_name = "AWS"
+}
+
+```
+
+## Argument Reference
+
+* `project_id` - (Required) The unique ID for the project to get all Cloud Provider Access 
+* `provider_name` - (Required) The cloud provider for which to create a new role. Currently only AWS is supported.
+* `iam_assumed_role_arn` - (Optional) - ARN of the IAM Role that Atlas assumes when accessing resources in your AWS account. This value is required after the creation (register of the role) as part of [Set Up Unified AWS Access](https://docs.atlas.mongodb.com/security/set-up-unified-aws-access/#set-up-unified-aws-access).
+
+
+## Attributes Reference
+
+* `id` - Unique identifier used by terraform for internal management.
+* `atlas_assumed_role_external_id` - Unique external ID Atlas uses when assuming the IAM role in your AWS account.
+* `atlas_aws_account_arn`          - ARN associated with the Atlas AWS account used to assume IAM roles in your AWS account.
+* `authorized_date`                - Date on which this role was authorized.
+* `created_date`                   - Date on which this role was created.
+* `feature_usages`                 - Atlas features this AWS IAM role is linked to.
+* `provider_name`                  - Name of the cloud provider. Currently limited to AWS.
+* `role_id`                        - Unique ID of this role.
+
+## Authorize role
+
+Once the resource is created add the field `iam_assumed_role_arn` see [Set Up Unified AWS Access](https://docs.atlas.mongodb.com/security/set-up-unified-aws-access/#set-up-unified-aws-access) , and execute a new `terraform apply` this will create a PATCH request.
```

Comment:
I wonder if there's any way to do this as one apply?   I guess without being able to do the security policy edit: https://docs.atlas.mongodb.com/security/set-up-unified-aws-access/#modify-your-aws-iam-role-trust-policy via an AWS terraform provider it has to be two steps?  Thoughts @leofigy and @shum ?

Reply:
I do think it has to be two separate steps given the need for the user to go their AWS console to set stuff up.

Reply:
Hi @themantissa, yea. the two steps are because 

- When the new role is register is in the mongodb atlas area in aws , these generates the new IAM ARN.
- To allow mongo to do things on-behalf of the customer account in aws, they need to create the policy associated the new ARNS (that were created by us in the first step)
- The authorize step in mongo, just validates that the relationship is complete. 

To have it in one apply, the issue is that we will ask the end-user to provide the aws credentials (through a profile)
that allows us to create the policy for them. I did this for the testing the authorization path (didn't added in this PR, It's is in a different branch). But I don't think is a good trade off, because we're importing the aws sdk. So we will be depending on them and we're assuming that the user has an account with the IAM roles to create policies, maybe if there is a demand from the users for a simplification we could add it as an optional flow.  But we must tell them explicitly the permissions they need, and we are creating a policy on their behalf like `mongo_access_<random>` or they add it as an input parameter.  

Sample code 

https://github.com/mongodb/terraform-provider-mongodbatlas/blob/c78edab8436113fcfc1ee0bfc0c29b120285b82f/mongodbatlas/resource_mongodbatlas_cloud_provider_access_test.go#L52

https://github.com/mongodb/terraform-provider-mongodbatlas/blob/c78edab8436113fcfc1ee0bfc0c29b120285b82f/mongodbatlas/resource_mongodbatlas_cloud_provider_access_test.go#L141

Reply:
@shum  and @leofigy thank you both so much.  @leofigy this is really really helpful and if it does come up as a pain point gives me just what I need to know.  Thank you!